zebra/rt_netlink.c

```diff
@@ -194,16 +194,22 @@ static inline int proto2zebra(int proto, int family)
 /*
 Pending: create an efficient table_id (in a tree/hash) based lookup)
  */
-static vrf_id_t vrf_lookup_by_table(u_int32_t table_id)
+static vrf_id_t vrf_lookup_by_table(u_int32_t table_id, ns_id_t ns_id)
 {
 	struct vrf *vrf;
 	struct zebra_vrf *zvrf;
 
 	RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
-		if ((zvrf = vrf->info) == NULL || (zvrf->table_id != table_id))
+		zvrf = vrf->info;
+		if (zvrf == NULL)
 			continue;
-
-		return zvrf_id(zvrf);
+		/* case vrf with netns : match the netnsid */
+		if (vrf_with_netns && ns_id == zvrf_id(zvrf))
+			return zvrf_id(zvrf);
+		if (zvrf->table_id != table_id)
+			continue;
+		if (!vrf_with_netns)
+			return zvrf_id(zvrf);
```

Comment:
This looks correct to me, but it would be a bit easier to follow if we rewrite it like this:
```c
static vrf_id_t vrf_lookup_by_table(u_int32_t table_id, ns_id_t ns_id)
{
        struct vrf *vrf;
        struct zebra_vrf *zvrf;

        RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
                zvrf = vrf->info;
                if (zvrf == NULL)
                        continue;

                switch (frr_vrf_backend) {
                case LINUX_VRF_DEV:
                        if (zvrf->table_id == table_id)
                                return vrf->vrf_id;
                        break;
                case LINUX_NETNS: 
                        if (vrf->vrf_id == ns_id)
                                return vrf->vrf_id;
                        break; 
                }
        }

        return VRF_DEFAULT;
}
```

Notice how the `frr_vrf_backend` enumeration replaces the `vrf_with_netns` variable. The code is now not only easier to read, but also easier to extend with new VRF backends in the future (e.g. OpenBSD rdomains).

Reply:
Another example of how this `frr_vrf_backend` enumeration can be useful:
```c
int vrf_create_socket(struct vrf *vrf, int domain, int type, int protocol)
{
        int fd;

        switch (frr_vrf_backend) {
        case LINUX_VRF_DEV:
                fd = socket(domain, type, protocol);
                if (fd == -1)
                        return -1;
                if (vrf->vrf_id != VRF_DEFAULT
                    && setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, vrf->name,
                                  strlen(vrf->name))
                               < 0) {
                        close(fd);
                        return -1;
                }
                break;
        case LINUX_NETNS:
                if (vrf_switch_to_netns(vrf) == -1)
                        return -1;
                fd = socket(domain, type, protocol);
                vrf_switchback_to_initial();
                break;
        case FREEBSD_VNET:
                if (freebsd_setfib(vrf) == -1)
                        return -1;
                fd = socket(domain, type, protocol);
                freebsd_setfib(default_vrf);
                break;
        case OPENBSD_RDOMAIN:
                /* TODO */
                break;
        }
}
```

We should try to abstract the internals of the VRF backends as much as possible in libfrr in order to avoid having platform-specific code throughout all the code.

Reply:
ok. I will rework a bit this part.